### PR TITLE
`Development`: Refactor ExamIntegrationTest

### PR DESCRIPTION
**Pull request by laadvo**
#### General

- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

The class `ExamIntegrationTest` has over 3000 lines of code and around 150 tests and can be considered to be a [blob](https://sourcemaking.com/antipatterns/the-blob). This antipattern also causes the integration test to be slow, since many attributes (sometimes only used in a few tests) to be initialized and the setup is expensive. In this PR I moved some of the tests, previously found in   `ExamIntegrationTest to new and already existing integration tests.

### Description

The following new classes were created:
- `ExamRegistrationIntegrationTest`: student registration for exams, and removing students from exams
- `ExamParticipationIntegrationTest`: testing related to participation
- `TestExamIntegrationTest`:  handels test exam related tests
- `ProgrammingExamIntegrationTest`: Houses tests directly linked to pro- gramming exercises incorporated within exams.
- `ExamStartTest`: Testing exam exercise start for different types of exercises.

Tests which fit into one of these classes have been moved, in `ExamIntegrationTest` remain exam creation, access, importing and other tests.

#### Code Review
- [x] Code Review 1
- [x] Code Review 2

### Test Coverage

unchanged

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/exam/ExamParticipationIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,1340 @@
...
+            repo.resetLocalRepo();
+        }
+
+        ParticipantScoreScheduleService.DEFAULT_WAITING_TIME_FOR_SCHEDULED_TASKS = 500;
+        participantScoreScheduleService.shutdown();
+    }
+
+    // TODO IMPORTANT test more complex exam configurations (mixed exercise type, more variants and more registered students)
+    @Nested
+    class ExamStartTest {
```
Comment (DominikRemo) [inline src/test/java/de/tum/in/www1/artemis/exam/ExamParticipationIntegrationTest.java:205]: ExamParticipationIntegrationTest is still a very long class. We should extract this nested class (```ExamStartTest```) into its own file :)

**Review by DominikRemo — COMMENTED** [review]
Test changes look mostly good to me. Left one small suggestion ^^

File: src/test/java/de/tum/in/www1/artemis/exam/ExamParticipationIntegrationTest.java (hunk 1 above)
Comment (laadvo) [reply src/test/java/de/tum/in/www1/artemis/exam/ExamParticipationIntegrationTest.java]: good idea, done ^^

**Review by DominikRemo — DISMISSED** [review]
Changes LGTM. Tests succeed on bamboo and GitHub Actions 👍
The number of tests on bamboo did not change, indicating that no test was lost

**Review by DominikRemo — APPROVED** [review]

File: src/test/java/de/tum/in/www1/artemis/exam/ExamIntegrationTest.java [hunk 2]
```diff
@@ -689,49 +209,6 @@ void testGenerateStudentExams() throws Exception {
...
-        assertThat(studentParticipations).isEmpty();
-
-        // invoke start exercises
-        studentExamService.startExercises(exam.getId()).join();
-
-        studentParticipations = participationTestRepository.findByExercise_ExerciseGroup_Exam_Id(exam.getId());
-        assertThat(studentParticipations).hasSize(12);
-
-        // Make sure delete also works if so many objects have been created before
-        request.delete("/api/courses/" + course1.getId() + "/exams/" + exam.getId(), HttpStatus.OK);
```
Comment (terlan98) [inline src/test/java/de/tum/in/www1/artemis/exam/ExamIntegrationTest.java:727]: This is missing in the corresponding function in `ExamParticipationIntegrationTest`. Is it on purpose?

File: src/test/java/de/tum/in/www1/artemis/exam/ExamParticipationIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,1169 @@
...
+        sourceGradingScale.setCourse(course);
+        gradingScaleRepository.save(sourceGradingScale);
+
+        var bonus = BonusFactory.generateBonus(bonusStrategy, -1.0, sourceGradingScale.getId(), bonusToGradingScale.getId());
+        bonusRepository.save(bonus);
+
+        course.setMaxPoints(100);
+        course.setPresentationScore(null);
+        courseRepo.save(course);
+
```
Comment (terlan98) [inline src/test/java/de/tum/in/www1/artemis/exam/ExamParticipationIntegrationTest.java:1145]:
```suggestion
```

File: src/test/java/de/tum/in/www1/artemis/exam/ProgrammingExamIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+        var programmingEx = programmingExerciseUtilService.addCourseExamExerciseGroupWithOneProgrammingExerciseAndTestCases();
+        var examWithProgrammingEx = programmingEx.getExerciseGroup().getExam();
+        examWithProgrammingEx.setVisibleDate(examWithProgrammingEx.getVisibleDate().plusSeconds(1));
+        request.put("/api/courses/" + examWithProgrammingEx.getCourse().getId() + "/exams", examWithProgrammingEx, HttpStatus.OK);
+        verify(instanceMessageSendService).sendProgrammingExerciseSchedule(programmingEx.getId());
+    }
+
+    @Test
+    @WithMockUser(username = TEST_PREFIX + "instructor1", roles = "INSTRUCTOR")
+    void testUpdateExam_reschedule_startDateChanged() throws Exception {
```
Comment (terlan98) [inline src/test/java/de/tum/in/www1/artemis/exam/ProgrammingExamIntegrationTest.java:145]: For consistency with other test names:
```suggestion
    void testUpdateExam_rescheduleProgramming_startDateChanged() throws Exception {
```

File: src/test/java/de/tum/in/www1/artemis/exam/TestExamIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    @Test
+    @WithMockUser(username = TEST_PREFIX + "instructor1", roles = "INSTRUCTOR")
+    void testCreateTestExam_asInstructor_badRequestWithWorkingTimeSetToZero() throws Exception {
+        // Test for bad request, if the working time is 0
+        Exam examD = ExamFactory.generateTestExam(course1);
+        examD.setWorkingTime(0);
+        request.post("/api/courses/" + course1.getId() + "/exams", examD, HttpStatus.BAD_REQUEST);
+
```
Comment (terlan98) [inline src/test/java/de/tum/in/www1/artemis/exam/TestExamIntegrationTest.java:143]:
```suggestion
```

File: src/test/java/de/tum/in/www1/artemis/exam/TestExamIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    @Test
+    @WithMockUser(username = TEST_PREFIX + "instructor1", roles = "INSTRUCTOR")
+    void testCreateTestExam_asInstructor_realExam_CorrectionRoundViolation() throws Exception {
+        Exam exam = ExamFactory.generateExam(course1);
+        exam.setNumberOfCorrectionRoundsInExam(0);
+        request.post("/api/courses/" + course1.getId() + "/exams", exam, HttpStatus.BAD_REQUEST);
+
+        exam.setNumberOfCorrectionRoundsInExam(3);
+        request.post("/api/courses/" + course1.getId() + "/exams", exam, HttpStatus.BAD_REQUEST);
+
```
Comment (terlan98) [inline src/test/java/de/tum/in/www1/artemis/exam/TestExamIntegrationTest.java:163]:
```suggestion
```

File: src/test/java/de/tum/in/www1/artemis/exam/ExamParticipationIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,1169 @@
...
+        assertThat(studentExamsDB).isEmpty();
+
+        // Fetch participations
+        exercises = examRepository.findAllExercisesByExamId(exam.getId()).toArray(new Exercise[0]);
+        participationList = new ArrayList<>();
+        for (Exercise exercise : exercises) {
+            participationList.addAll(studentParticipationRepository.findByExerciseId(exercise.getId()));
+        }
+        assertThat(participationList).hasSize(12);
+
```
Comment (terlan98) [inline src/test/java/de/tum/in/www1/artemis/exam/ExamParticipationIntegrationTest.java:236]:
```suggestion
```

**Review by terlan98 — CHANGES_REQUESTED** [review]
Changes look good in general 👍 Left some small comments/suggestions:

- Shouldn't it be moved to `ExamRegistrationIntegrationTest`? https://github.com/ls1intum/Artemis/blob/69cb7cd2979bf049bc53077cb88aea8819a0fdf9/src/test/java/de/tum/in/www1/artemis/exam/ExamIntegrationTest.java#L755-L760

File: src/test/java/de/tum/in/www1/artemis/exam/ExamIntegrationTest.java (hunk 2 above)
Comment (laadvo) [reply src/test/java/de/tum/in/www1/artemis/exam/ExamIntegrationTest.java:722]: good catch, it was by accident :sweat_smile:

File: src/test/java/de/tum/in/www1/artemis/exam/ProgrammingExamIntegrationTest.java (hunk 4 above)
Comment (laadvo) [reply src/test/java/de/tum/in/www1/artemis/exam/ProgrammingExamIntegrationTest.java]: done

**Review by terlan98 — APPROVED** [review]
Code looks good. Thanks for the changes 👍

**Review by DominikRemo — APPROVED** [review]
LGTM 👍

**Pull request merged**